source/module_hamilt_lcao/module_gint/cal_ddpsir_ylm.cpp

```diff
@@ -1,52 +1,62 @@
 #include "gint_tools.h"
+#include "module_base/global_function.h"
 #include "module_base/timer.h"
 #include "module_base/ylm.h"
 namespace Gint_Tools{
 void cal_ddpsir_ylm(
-    const Grid_Technique& gt, const int bxyz,
+    const Grid_Technique& gt, 
+    const int bxyz,
     const int na_grid,                 // number of atoms on this grid
     const int grid_index,              // 1d index of FFT index (i,j,k)
     const double delta_r,              // delta_r of the uniform FFT grid
     const int* const block_index,      // block_index[na_grid+1], count total number of atomis orbitals
     const int* const block_size,       // block_size[na_grid],	number of columns of a band
     const bool* const* const cal_flag, // cal_flag[bxyz][na_grid],	whether the atom-grid distance is larger than cutoff
-    double* const* const ddpsir_ylm_xx, double* const* const ddpsir_ylm_xy, double* const* const ddpsir_ylm_xz,
-    double* const* const ddpsir_ylm_yy, double* const* const ddpsir_ylm_yz, double* const* const ddpsir_ylm_zz)
+    double* const* const ddpsir_ylm_xx, 
+    double* const* const ddpsir_ylm_xy, 
+    double* const* const ddpsir_ylm_xz,
+    double* const* const ddpsir_ylm_yy, 
+    double* const* const ddpsir_ylm_yz, 
+    double* const* const ddpsir_ylm_zz)
 {
     ModuleBase::timer::tick("Gint_Tools", "cal_ddpsir_ylm");
+
+    int it;
+    double distance;
+    double dr[3];
+    double mt[3];
+    Atom* atom;
     const UnitCell& ucell = *gt.ucell;
     std::vector<const double*> it_psi_uniform(gt.nwmax);
     std::vector<const double*> it_dpsi_uniform(gt.nwmax);
     std::vector<const double*> it_d2psi_uniform(gt.nwmax);
-    std::vector<int> it_psi_nr_uniform(gt.nwmax);
+
     // array to store spherical harmonics and its derivatives
     // the first dimension equals 36 because the maximum nwl is 5.
     double rly[36];
     ModuleBase::Array_Pool<double> grly(36, 3);
-
+    ModuleBase::Array_Pool<double*> dpsi(gt.nwmax,6);
+    for (int i=0; i<dpsi.get_nc()*dpsi.get_nr(); i++)
+    {
+        dpsi.get_ptr_1D()[i] = new double[3];
```

Comment:
could you use std::vector instead?

Reply:
It might be more appropriate to use std::array instead. Here, a fixed-length array is used, and std::array balances efficiency and memory management.

Reply:
The array can also be modified to pybind format, while also considering Pythonization.